drafts/npm.md

```diff
@@ -0,0 +1,472 @@
+may generate large pull requests in the repository's commit history.
+
+#### Hash pinning
+
+Hash pinning informs the package manager the expected hash for each dependency,
+without trusting the registries. The package manager then verifies, during each
+installation, that the hash of each dependency remains the same.
+
+Npm provides two options to achieve hash pinning.
+
+##### Package-lock.json
+
+The
+[package-lock.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json)
+contains ***all*** the dependencies (direct and indirect) along with a
+cryptographic hash of their content:
+
+```json
+{
+  "name": "A",
+  "version": "0.1.0",
+  ...metadata fields...
+  "dependencies": {
+    "B": {
+      "version": "0.0.1",
+      "resolved": "https://registry.npmjs.org/B/-/B-0.0.1.tgz",
+      "integrity": "sha512-DeAdb33F+"
+      "dependencies": {
+        "C": {
+          "version": "git://github.com/org/C.git#5c380ae319fc4efe9e7f2d9c78b0faa588fd99b4"
+        }
+      }
+    }
+  }
+}
+```
+
+The package-lock.json file is a ***snapshot of an installation*** that allows
+later reproduction of the same installation. As such, the lock file is
+generated or updated via the various commands that install packages, e.g., `npm
+install`. If some dependencies are missing or not pinned by hash (e.g.,
+`integrity` field is not present), the installation will patch the lock file to
+reflect the installation.
+
+The lock file ***cannot*** be uploaded to a registry, which means that
+consumers who locally install the package via `npm install` may [see different
+dependency
+versions](https://dev.to/saurabhdaware/but-what-the-hell-is-package-lock-json-b04)
+than the repo owners used during testing. Using package-lock.json is akin to
+dynamic linking for low-level programming languages: the loader will resolve
+dependencies at load time using libraries available on the system. Using this
+lock file leaves the task of deciding dependencies' versions to use to the
+package consumer.
+
+##### Shrinkwrap.json
+
+[Shrinkwrap.json](https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#package-lockjson-vs-npm-shrinkwrapjson)
+is another lock file supported by npm. The main difference is that this lock
+file ***may*** be uploaded to a registry along with the package. This ensures that
+consumers of the dependency will obtain the same dependencies' hashes as the
+repo owners intended. With shrinkwrap.json, the package producer takes
+responsibility for updating the dependencies on behalf of their consumers. It's
+akin to static linking for low-level programming languages: everything is
+declared at packaging time.
+
+To generate the shrinkwrap.json, an existing package-lock.json must be present
+and the command [`npm
+shrinkwrap`](https://docs.npmjs.com/cli/v8/commands/npm-shrinkwrap) must be
+run.
+
+**Recommendations:**
+
+1. Developers should declare and commit a manifest file for ***all*** their
```

Comment:
i strongly disagree with this recommendation, for the reasons given above, and none of my 300+ packages (which represent well over 10% of npm's download traffic) include a lockfile, nor will they, because I think that would cause far more harm than it helps.

Reply:
here we don't mean shrinkwrap files, we just mean a lock file used for _CI_ - which is not used during package installation and won't "blocks the free flow of bugfixes and nonbreaking updates". 

I agree the way it's written is confusing. Thanks for pointing this out.

Reply:
Yes, but I still think that such a dev-only lockfile is harmful, because it falsely insulates maintainers from problems their users will encounter.